diff --git i/dlls/d2d1/geometry.c w/dlls/d2d1/geometry.c
index 633919af..fe5ef160 100644
--- i/dlls/d2d1/geometry.c
+++ w/dlls/d2d1/geometry.c
@@ -972,6 +972,230 @@ static bool d2d_figure_add_lines(struct d2d_figure *figure, const D2D1_POINT_2F
     return true;
 }
 
+static bool d2d_arc_check_radius(float halfchord2, float fuzz2, float *radius)
+{
+    bool accept = !(*radius * *radius <= halfchord2 * fuzz2);
+    if (accept)
+    {
+        if (*radius < 0.0f)
+            *radius = -*radius;
+    }
+    return accept;
+}
+
+static int d2d_arc_get_piece_count(const D2D_POINT_2F *start, const D2D_POINT_2F *end,
+        bool large_arc, bool sweep_up, float *cos_angle, float *sin_angle)
+{
+    float angle;
+    int count;
+
+    *cos_angle = d2d_point_dot(start, end);
+    *sin_angle = start->x * end->y - start->y * end->x;
+
+    if (*cos_angle >= 0.0f)
+    {
+        if (!large_arc) return 1;
+        count = 4;
+    }
+    else
+    {
+        count = large_arc ? 3 : 2;
+    }
+
+    angle = atan2f(*sin_angle, *cos_angle);
+    if (sweep_up)
+    {
+        if (angle < 0.0f)
+            angle += 2.0f * M_PI;
+    }
+    else
+    {
+        if (angle > 0.0f)
+            angle -= 2.0f * M_PI;
+    }
+
+    angle /= count;
+    *cos_angle = cosf(angle);
+    *sin_angle = sinf(angle);
+
+    return count;
+}
+
+static float d2d_arc_get_bezier_distance(float rDot, bool sweep_up)
+{
+    float denom_squared, denominator;
+    const float fuzz = 1.e-6;
+    float numerator, rA, dist;
+
+    rA = 0.5f * (1.0f + rDot);
+    if (rA < 0.0f)
+        return 0.0f;
+
+    denom_squared = 1.0f - rA;
+    if (denom_squared <= 0.0f)
+        return 0.0f;
+
+    denominator = sqrt(denom_squared);
+    numerator = (4.0f / 3.0f) * (1.0f - sqrt(rA));
+
+    dist = (numerator <= denominator * fuzz) ? 0.0f : numerator / denominator;
+    if (!sweep_up)
+        dist = -dist;
+
+    return dist;
+}
+
+/* Approximation logic is taken in its entirety from WpfGfx graphics core of
+   Windows Presentation Foundation framework, distributed under MIT license. */
+static int d2d_arc_to_bezier(const D2D_POINT_2F *start_point, const D2D1_ARC_SEGMENT *arc,
+        D2D_POINT_2F *points)
+{
+    float x, y, rHalfChord2, rCos, rSin, rCosArcAngle, rSinArcAngle, dist, rotation;
+    D2D_POINT_2F ptStart, ptEnd, vecToBez1, vecToBez2;
+    const float FUZZ = 1.e-6;
+    const float PI_OVER_180 = 0.0174532925199432957692;
+    bool large_arc, sweep_up, zero_center = false;
+    D2D_POINT_2F center, radius;
+    D2D1_MATRIX_3X2_F m;
+    int cPieces = -1;
+
+    float fuzz2 = FUZZ * FUZZ;
+    int i, j;
+
+    d2d_point_set(&radius, arc->size.width, arc->size.height);
+    rotation = arc->rotationAngle;
+    large_arc = arc->arcSize == D2D1_ARC_SIZE_LARGE;
+    sweep_up = arc->sweepDirection == D2D1_SWEEP_DIRECTION_CLOCKWISE;
+
+    x = 0.5f * (arc->point.x - start_point->x);
+    y = 0.5f * (arc->point.y - start_point->y);
+
+    rHalfChord2 = x * x + y * y;
+    if (rHalfChord2 < fuzz2)
+        return -1;
+
+    if (!d2d_arc_check_radius(rHalfChord2, fuzz2, &radius.x) ||
+        !d2d_arc_check_radius(rHalfChord2, fuzz2, &radius.y))
+    {
+        points[0] = arc->point;
+        return 0;
+    }
+
+    if (fabs(rotation) < FUZZ)
+    {
+        rCos = 1.0f;
+        rSin = 0.0f;
+    }
+    else
+    {
+        float tmp;
+
+        rotation = -rotation * PI_OVER_180;
+
+        rCos = cosf(rotation);
+        rSin = sinf(rotation);
+
+        tmp = x * rCos - y * rSin;
+        y = x * rSin + y * rCos;
+        x = tmp;
+    }
+
+    x /= radius.x;
+    y /= radius.y;
+
+    rHalfChord2 = x * x + y * y;
+    if (rHalfChord2 > 1.0f)
+    {
+        float tmp = sqrtf(rHalfChord2);
+
+        radius.x *= tmp;
+        radius.y *= tmp;
+        center.x = center.y = 0.0f;
+        zero_center = true;
+
+        x /= tmp;
+        y /= tmp;
+    }
+    else
+    {
+        float tmp = sqrtf((1.0f - rHalfChord2) / rHalfChord2);
+        if (large_arc != sweep_up)
+        {
+            center.x = -tmp * y;
+            center.y = tmp * x;
+        }
+        else
+        {
+            center.x = tmp * y;
+            center.y = -tmp * x;
+        }
+    }
+
+    d2d_point_set(&ptStart, -x - center.x, -y - center.y);
+    d2d_point_set(&ptEnd, x - center.x, y - center.y);
+
+    m._11 = rCos * radius.x;
+    m._12 = -rSin * radius.x;
+    m._21 = rSin * radius.y;
+    m._22 = rCos * radius.y;
+    m._31 = 0.5f * (arc->point.x + start_point->x);
+    m._32 = 0.5f * (arc->point.y + start_point->y);
+    if (!zero_center)
+    {
+        m._31 += (m._11 * center.x + m._12 * center.y);
+        m._32 += (m._21 * center.x + m._22 * center.y);
+    }
+
+    cPieces = d2d_arc_get_piece_count(&ptStart, &ptEnd, large_arc, sweep_up, &rCosArcAngle, &rSinArcAngle);
+
+    dist = d2d_arc_get_bezier_distance(rCosArcAngle, sweep_up);
+    d2d_point_set(&vecToBez1, -dist * ptStart.y, dist * ptStart.x);
+
+    j = 0;
+    for (i = 1; i < cPieces; ++i)
+    {
+        D2D_POINT_2F ptPieceEnd;
+
+        d2d_point_set(&ptPieceEnd, ptStart.x * rCosArcAngle - ptStart.y * rSinArcAngle,
+                              ptStart.x * rSinArcAngle + ptStart.y * rCosArcAngle);
+        d2d_point_set(&vecToBez2, -dist * ptPieceEnd.y, dist * ptPieceEnd.x);
+
+        d2d_point_transform(&points[j++], &m, ptStart.x + vecToBez1.x, ptStart.y + vecToBez1.y);
+        d2d_point_transform(&points[j++], &m, ptPieceEnd.x - vecToBez2.x, ptPieceEnd.y - vecToBez2.y);
+        d2d_point_transform(&points[j++], &m, ptPieceEnd.x, ptPieceEnd.y);
+
+        ptStart = ptPieceEnd;
+        vecToBez1 = vecToBez2;
+    }
+
+    d2d_point_set(&vecToBez2, -dist * ptEnd.y, dist * ptEnd.x);
+    d2d_point_transform(&points[j++], &m, ptStart.x + vecToBez1.x, ptStart.y + vecToBez1.y);
+    d2d_point_transform(&points[j++], &m, ptEnd.x - vecToBez2.x, ptEnd.y - vecToBez2.y);
+    d2d_point_set(&points[j], arc->point.x, arc->point.y);
+
+    return cPieces;
+}
+
+static bool d2d_figure_add_arc(struct d2d_figure *figure, const D2D1_ARC_SEGMENT *arc)
+{
+    size_t last = figure->vertex_count - 1;
+    D2D_POINT_2F points[12];
+    int count = 0;
+
+    count = d2d_arc_to_bezier(&figure->vertices[last], arc, points);
+
+    if (count > 0)
+    {
+        return d2d_figure_add_beziers(figure, (D2D1_BEZIER_SEGMENT *)points, count);
+    }
+    else if (count == 0)
+    {
+        return d2d_figure_add_lines(figure, points, 1);
+    }
+
+    return true;
+}
+
 static void d2d_figure_cleanup(struct d2d_figure *figure)
 {
     free(figure->original_bezier_controls);
@@ -3363,7 +3587,7 @@ static void STDMETHODCALLTYPE d2d_geometry_sink_AddArc(ID2D1GeometrySink *iface,
 {
     struct d2d_geometry *geometry = impl_from_ID2D1GeometrySink(iface);
 
-    FIXME("iface %p, arc %p stub!\n", iface, arc);
+    TRACE("iface %p, arc %p.\n", iface, arc);
 
     if (geometry->u.path.state != D2D_GEOMETRY_STATE_FIGURE)
     {
@@ -3371,9 +3595,9 @@ static void STDMETHODCALLTYPE d2d_geometry_sink_AddArc(ID2D1GeometrySink *iface,
         return;
     }
 
-    if (!d2d_figure_add_vertex(&geometry->u.path.figures[geometry->u.path.figure_count - 1], arc->point))
+    if (!d2d_figure_add_arc(&geometry->u.path.figures[geometry->u.path.figure_count - 1], arc))
     {
-        ERR("Failed to add vertex.\n");
+        d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
         return;
     }
 
@@ -5814,8 +6038,13 @@ static const struct ID2D1GeometryGroupVtbl d2d_geometry_group_vtbl =
 HRESULT d2d_geometry_group_init(struct d2d_geometry *geometry, ID2D1Factory *factory,
         D2D1_FILL_MODE fill_mode, ID2D1Geometry **geometries, unsigned int geometry_count)
 {
-    unsigned int i;
+    unsigned int i, j;
+    struct d2d_geometry *other_geom;
+    D2D_MATRIX_3X2_F g, gplain;
+    size_t f_vertex_count, f_face_count, f_bezier_vertex_count, f_arc_vertex_count;
+    size_t o_vertex_count, o_face_count, o_bezier_count, o_bezier_face_count, o_arc_count, o_arc_face_count;
 
+    FIXME("Ignoring fill_mode=%#x!\n", fill_mode);
     d2d_geometry_init(geometry, factory, &identity, (ID2D1GeometryVtbl *)&d2d_geometry_group_vtbl);
 
     if (!(geometry->u.group.src_geometries = calloc(geometry_count, sizeof(*geometries))))
@@ -5824,13 +6053,242 @@ HRESULT d2d_geometry_group_init(struct d2d_geometry *geometry, ID2D1Factory *fac
         return E_OUTOFMEMORY;
     }
 
+    geometry->fill.vertex_count = 0;
+    geometry->fill.face_count = 0;
+    geometry->fill.bezier_vertex_count = 0;
+    geometry->fill.arc_vertex_count = 0;
+    geometry->outline.vertex_count = 0;
+    geometry->outline.face_count = 0;
+    geometry->outline.bezier_count = 0;
+    geometry->outline.bezier_face_count = 0;
+    geometry->outline.arc_count = 0;
+    geometry->outline.arc_face_count = 0;
     for (i = 0; i < geometry_count; ++i)
     {
         ID2D1Geometry_AddRef(geometry->u.group.src_geometries[i] = geometries[i]);
+        other_geom = unsafe_impl_from_ID2D1Geometry(geometries[i]);
+        geometry->fill.vertex_count += other_geom->fill.vertex_count;
+        geometry->fill.face_count += other_geom->fill.face_count;
+        geometry->fill.bezier_vertex_count += other_geom->fill.bezier_vertex_count;
+        geometry->fill.arc_vertex_count += other_geom->fill.arc_vertex_count;
+        geometry->outline.vertex_count += other_geom->outline.vertex_count;
+        geometry->outline.face_count += other_geom->outline.face_count;
+        geometry->outline.bezier_count += other_geom->outline.bezier_count;
+        geometry->outline.bezier_face_count += other_geom->outline.bezier_face_count;
+        geometry->outline.arc_count += other_geom->outline.arc_count;
+        geometry->outline.arc_face_count += other_geom->outline.arc_face_count;
     }
     geometry->u.group.geometry_count = geometry_count;
     geometry->u.group.fill_mode = fill_mode;
 
+    if (!(geometry->fill.vertices = calloc(geometry->fill.vertex_count, sizeof(D2D1_POINT_2F))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    if (!(geometry->fill.faces = calloc(geometry->fill.face_count, sizeof(struct d2d_face))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->fill.faces_size = geometry->fill.face_count;
+    if (!(geometry->fill.bezier_vertices = calloc(geometry->fill.bezier_vertex_count, sizeof(struct d2d_curve_vertex))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->fill.bezier_vertices_size = geometry->fill.bezier_vertex_count;
+    if (!(geometry->fill.arc_vertices = calloc(geometry->fill.arc_vertex_count, sizeof(struct d2d_curve_vertex))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->fill.arc_vertices_size = geometry->fill.arc_vertex_count;
+    if (!(geometry->outline.vertices = calloc(geometry->outline.vertex_count, sizeof(struct d2d_outline_vertex))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->outline.vertices_size = geometry->outline.vertex_count;
+    if (!(geometry->outline.faces = calloc(geometry->outline.face_count, sizeof(struct d2d_face))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->outline.faces_size = geometry->outline.face_count;
+    if (!(geometry->outline.beziers = calloc(geometry->outline.bezier_count, sizeof(struct d2d_curve_outline_vertex))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->outline.beziers_size = geometry->outline.bezier_count;
+    if (!(geometry->outline.bezier_faces = calloc(geometry->outline.bezier_face_count, sizeof(struct d2d_face))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->outline.bezier_faces_size = geometry->outline.bezier_face_count;
+    if (!(geometry->outline.arcs = calloc(geometry->outline.arc_count, sizeof(struct d2d_curve_outline_vertex))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->outline.arcs_size = geometry->outline.arc_count;
+    if (!(geometry->outline.arc_faces = calloc(geometry->outline.arc_face_count, sizeof(struct d2d_face))))
+    {
+        for (i = 0; i < geometry->u.group.geometry_count; ++i)
+            ID2D1Geometry_Release(geometry->u.group.src_geometries[i]);
+        d2d_geometry_cleanup(geometry);
+        return E_OUTOFMEMORY;
+    }
+    geometry->outline.arc_faces_size = geometry->outline.arc_face_count;
+
+    f_vertex_count = 0;
+    f_face_count = 0;
+    f_bezier_vertex_count = 0;
+    f_arc_vertex_count = 0;
+    o_vertex_count = 0;
+    o_face_count = 0;
+    o_bezier_count = 0;
+    o_bezier_face_count = 0;
+    o_arc_count = 0;
+    o_arc_face_count = 0;
+    for (i = 0; i < geometry_count; ++i)
+    {
+        other_geom = unsafe_impl_from_ID2D1Geometry(geometries[i]);
+        g = other_geom->transform;
+        gplain = g;
+        gplain.dx = 0.0f;
+        gplain.dy = 0.0f;
+
+        for (j = 0; j < other_geom->fill.vertex_count; ++j) {
+            d2d_point_transform(&geometry->fill.vertices[j+f_vertex_count], &g,
+                    other_geom->fill.vertices[j].x, other_geom->fill.vertices[j].y);
+        }
+        for (j = 0; j < other_geom->fill.face_count; ++j) {
+            geometry->fill.faces[f_face_count+j].v[0] = other_geom->fill.faces[j].v[0]
+                + (UINT16) f_vertex_count;
+            geometry->fill.faces[f_face_count+j].v[1] = other_geom->fill.faces[j].v[1]
+                + (UINT16) f_vertex_count;
+            geometry->fill.faces[f_face_count+j].v[2] = other_geom->fill.faces[j].v[2]
+                + (UINT16) f_vertex_count;
+        }
+        f_vertex_count += other_geom->fill.vertex_count;
+        f_face_count += other_geom->fill.face_count;
+
+        for (j = 0; j < other_geom->fill.bezier_vertex_count; ++j) {
+            d2d_point_transform(&geometry->fill.bezier_vertices[j+f_bezier_vertex_count].position,
+                    &g, other_geom->fill.bezier_vertices[j].position.x,
+                    other_geom->fill.bezier_vertices[j].position.y);
+            geometry->fill.bezier_vertices[j+f_bezier_vertex_count].texcoord
+                = other_geom->fill.bezier_vertices[j].texcoord;
+        }
+        f_bezier_vertex_count += other_geom->fill.bezier_vertex_count;
+
+        for (j = 0; j < other_geom->fill.arc_vertex_count; ++j) {
+            d2d_point_transform(&geometry->fill.arc_vertices[j+f_arc_vertex_count].position,
+                    &g, other_geom->fill.arc_vertices[j].position.x,
+                    other_geom->fill.arc_vertices[j].position.y);
+            geometry->fill.arc_vertices[j+f_arc_vertex_count].texcoord
+                = other_geom->fill.arc_vertices[j].texcoord;
+
+        }
+        f_arc_vertex_count += other_geom->fill.arc_vertex_count;
+
+        for (j = 0; j < other_geom->outline.vertex_count; ++j) {
+            d2d_point_transform(&geometry->outline.vertices[j+o_vertex_count].position, &g,
+                    other_geom->outline.vertices[j].position.x, other_geom->outline.vertices[j].position.y);
+            d2d_point_transform(&geometry->outline.vertices[j+o_vertex_count].prev, &gplain,
+                    other_geom->outline.vertices[j].prev.x, other_geom->outline.vertices[j].prev.y);
+            d2d_point_normalise(&geometry->outline.vertices[j+o_vertex_count].prev);
+            d2d_point_transform(&geometry->outline.vertices[j+o_vertex_count].next, &gplain,
+                    other_geom->outline.vertices[j].next.x, other_geom->outline.vertices[j].next.y);
+            d2d_point_normalise(&geometry->outline.vertices[j+o_vertex_count].next);
+        }
+        for (j = 0; j < other_geom->outline.face_count; ++j) {
+            geometry->outline.faces[o_face_count+j].v[0] = other_geom->outline.faces[j].v[0]
+                + (UINT16) o_vertex_count;
+            geometry->outline.faces[o_face_count+j].v[1] = other_geom->outline.faces[j].v[1]
+                + (UINT16) o_vertex_count;
+            geometry->outline.faces[o_face_count+j].v[2] = other_geom->outline.faces[j].v[2]
+                + (UINT16) o_vertex_count;
+        }
+        o_vertex_count += other_geom->outline.vertex_count;
+        o_face_count += other_geom->outline.face_count;
+
+        for (j = 0; j < other_geom->outline.bezier_count; ++j) {
+            d2d_point_transform(&geometry->outline.beziers[j+o_bezier_count].position, &g,
+                    other_geom->outline.beziers[j].position.x, other_geom->outline.beziers[j].position.y);
+            d2d_point_transform(&geometry->outline.beziers[j+o_bezier_count].prev, &gplain,
+                    other_geom->outline.beziers[j].prev.x, other_geom->outline.beziers[j].prev.y);
+            d2d_point_normalise(&geometry->outline.beziers[j+o_bezier_count].prev);
+            d2d_point_transform(&geometry->outline.beziers[j+o_bezier_count].next, &gplain,
+                    other_geom->outline.beziers[j].next.x, other_geom->outline.beziers[j].next.y);
+            d2d_point_normalise(&geometry->outline.beziers[j+o_bezier_count].next);
+            d2d_point_transform(&geometry->outline.beziers[j+o_bezier_count].p0, &g,
+                    other_geom->outline.beziers[j].p0.x, other_geom->outline.beziers[j].p0.y);
+            d2d_point_transform(&geometry->outline.beziers[j+o_bezier_count].p1, &g,
+                    other_geom->outline.beziers[j].p1.x, other_geom->outline.beziers[j].p1.y);
+            d2d_point_transform(&geometry->outline.beziers[j+o_bezier_count].p2, &g,
+                    other_geom->outline.beziers[j].p2.x, other_geom->outline.beziers[j].p2.y);
+        }
+        for (j = 0; j < other_geom->outline.bezier_face_count; ++j) {
+            geometry->outline.bezier_faces[o_bezier_face_count+j].v[0]
+                = other_geom->outline.bezier_faces[j].v[0] + (UINT16) o_bezier_count;
+            geometry->outline.bezier_faces[o_bezier_face_count+j].v[1]
+                = other_geom->outline.bezier_faces[j].v[1] + (UINT16) o_bezier_count;
+            geometry->outline.bezier_faces[o_bezier_face_count+j].v[2]
+                = other_geom->outline.bezier_faces[j].v[2] + (UINT16) o_bezier_count;
+        }
+        o_bezier_count += other_geom->outline.bezier_count;
+        o_bezier_face_count += other_geom->outline.bezier_face_count;
+
+        for (j = 0; j < other_geom->outline.arc_count; ++j) {
+            d2d_point_transform(&geometry->outline.arcs[j+o_arc_count].position, &g,
+                    other_geom->outline.arcs[j].position.x, other_geom->outline.arcs[j].position.y);
+            d2d_point_transform(&geometry->outline.arcs[j+o_arc_count].prev, &gplain,
+                    other_geom->outline.arcs[j].prev.x, other_geom->outline.arcs[j].prev.y);
+            d2d_point_normalise(&geometry->outline.arcs[j+o_arc_count].prev);
+            d2d_point_transform(&geometry->outline.arcs[j+o_arc_count].next, &gplain,
+                    other_geom->outline.arcs[j].next.x, other_geom->outline.arcs[j].next.y);
+            d2d_point_normalise(&geometry->outline.arcs[j+o_arc_count].next);
+            d2d_point_transform(&geometry->outline.arcs[j+o_arc_count].p0, &g,
+                    other_geom->outline.arcs[j].p0.x, other_geom->outline.arcs[j].p0.y);
+            d2d_point_transform(&geometry->outline.arcs[j+o_arc_count].p1, &g,
+                    other_geom->outline.arcs[j].p1.x, other_geom->outline.arcs[j].p1.y);
+            d2d_point_transform(&geometry->outline.arcs[j+o_arc_count].p2, &g,
+                    other_geom->outline.arcs[j].p2.x, other_geom->outline.arcs[j].p2.y);
+        }
+        for (j = 0; j < other_geom->outline.arc_face_count; ++j) {
+            geometry->outline.arc_faces[o_arc_face_count+j].v[0]
+                = other_geom->outline.arc_faces[j].v[0] + (UINT16) o_arc_count;
+            geometry->outline.arc_faces[o_arc_face_count+j].v[1]
+                = other_geom->outline.arc_faces[j].v[1] + (UINT16) o_arc_count;
+            geometry->outline.arc_faces[o_arc_face_count+j].v[2]
+                = other_geom->outline.arc_faces[j].v[2] + (UINT16) o_arc_count;
+        }
+        o_arc_count += other_geom->outline.arc_count;
+        o_arc_face_count += other_geom->outline.arc_face_count;
+    }
+
     return S_OK;
 }
 
